nbgrader/apps/instantiatetestsapp.py

```diff
@@ -0,0 +1,107 @@
+# coding: utf-8
```

Comment:
I am a bit hesitant about adding a new command (especially one that requires creating a new directory in the nbgrader directory) as that will increase complexity and make things more confusing for folks new to nbgrader. But I guess the reason is so you can run validation on the source assignment with the actual tests that will be used (because if you run validation on the notebooks in `source`, they don't yet have the tests, and if you run validation in `release`, they don't have the solutions). Hmm.

What would you think about the following instead: rather than a new top-level command, we add an option to `GenerateAssignmentApp` (something like `nbgrader generate_assignment --instantiate_tests`) that will, as a byproduct of generating the assignment, save the intermediate notebooks to `source_with_tests/` (rather than `instantiated/`). Then the functionality is there to generate the notebooks if you want to get them with both the solutions and the hidden tests, but it's more of a debugging tool rather than a key step in the nbgrader process. I think it should be a pretty simple change, I think you can just add the following to `generateassignmentsapp.py` before running the generate converter (plus code to implement the option):

```
if self.instantiate_tests:
    converter = InstantiateTests(coursedir=self.coursedir, parent=self)
    try:
        converter.start()
    except NbGraderException:
        sys.exit(1)
```

Reply:
@jhamrick thanks for the review! We'll start working on these suggestions shortly. One clarification/question before we get started:

> I am a bit hesitant about adding a new command (especially one that requires creating a new directory in the nbgrader directory) as that will increase complexity and make things more confusing for folks new to nbgrader. But I guess the reason is so you can run validation on the source assignment with the actual tests that will be used (because if you run validation on the notebooks in source, they don't yet have the tests, and if you run validation in release, they don't have the solutions)

The `instantiate_tests` converter is not a required part of the workflow. Users can simply use nbgrader as they would right now (run `generate_assignment` to go from `source/` to `release/`). The `instantiate_tests` converter is optional -- just in case the instructor wants to play with an editable version of the assignment that has both solutions and test code in it.

I think I'd personally be confused that `generate_assignment --instantiate_tests` doesn't produce a `release/` version. I think I'd also assume (before digging into docs) that adding the `--instantiate_tests` flag forces autotest to run (and that it won't run without that flag), which isn't the case. 

Semi-related: I am happy to change `instantiated/` to `source_with_tests/`. But it is a bit unwieldy, and if someone is running the `instantiate_tests` converter (or `--instantiate_tests` flag), the name `instantiated/` would hopefully not be confusing... I don't feel too strongly about it though.

Thoughts? Let us know how you'd like to proceed and we'll get to it. I skimmed your other comments and I agree with all!

Reply:
(oops -- we actually had one more question regarding adding a new example notebook to the docs! See above :) )

Reply:
> The instantiate_tests command is not a required part of the workflow.

Yes, I understand, it's just my gut feeling that making it a top-level command implies that it is (though I guess who knows, that's really an empirical question!)

> I think I'd personally be confused that generate_assignment --instantiate_tests doesn't produce a release/ version.

Ah so my thinking was that it would *both* generate a `release/` version and a `source_with_tests/` version. In other words, because making the tests is an intermediate step of `generate_assignment`, adding the flag tells nbgrader to save out that intermediate step separately. Maybe it would be clearer if, instead of calling the flag `--instantiate_tests`, we called it `--generate_source_with_tests` or something like that? And then that would address your other point about the name `source_with_tests/` being a little confusing. What do you think?

Reply:
Ah -- generating both a `release/` and `source_with_tests/` (given the `--blah` flag) seems like reasonable behaviour to me!

But looking at the code again, I'm not sure how easy it would be to modify the `generate_assignment` converter to output two copies, with one only using a partial set of preprocessors. It seems the base converter assumes one input notebook / folder -> one output notebook / folder. I could be wrong here -- maybe you know a convenient way to pull that off. I think unless there's a clean way of doing that (without hacking the `generate_assignments` converter substantially) I'd rather keep it as a separate converter that follows existing code patterns.

(I'm of course happy to add documentation to make it super clear that the `instantiate_tests` converter is optional and what its purpose is, in case we go that route)

Reply:
(if we keep the separate converter, would you prefer `nbgrader generate_source_with_tests` or `nbgrader instantiate_tests`?)

Reply:
So it wouldn't be a change to the converter itself, but to the application. I think the code I posted above that modifies `generateassignmentsapp.py` should do the trick---the idea is basically just to run both the instantiate tests converter *and* the generate assignmemt converter.

(Technically, this would cause the instantiate tests preprocessor to run twice, but I don't think the overhead should be a problem)

Reply:
I dig it -- happy to go with that solution then! We'll get to it.

Reply:
We added a new flag called --generate_source_with_tests to the `generateassignmentsapp` and removed the other app